### ISSUE TITLE
[BUG] Accessibility - App-  Horizontal scrolling is not accessible (multiple areas of app affected by this)

### ISSUE THREAD

**Issue by acovher2** [rokwire/illinois-app#2093]
**Describe the accessibility issue**
Currently any horizontally scrolling widget is in violation. We need a way for non-sighted users to be able to know there is content to swipe to.

 **Summary of requirements for Next/previous controls:**

- Controls should be separate from the individual twitter cards, (Perhaps at top of feed as in example from Precision planting)
- Controls must have disabled & enabled state
- Controls must have a pressed state
- Controls must have a focus state
- Controls must meet the Color contrast requirements for [WCAG 1.1.1 Non-text contrast](https://www.w3.org/TR/WCAG21/#dfn-non-text-content) and [WCAG 1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast)

**To Reproduce
Steps to reproduce the behavior:**

1. Open app, see horizontally scrolling widgets. 
2. Attempt to use a screen reader like VoiceOVer or Talkback and gesture through them.... you can't. 

**Expected behavior**
There should be Previous/Next controls for any horizontally scrolling widget. 
These should not be located on each individual card. 
These Previous/next controls also need disabled/enabled states, as well as Pressed and Focus states.
The controls should meet color contrast ratio

**Screenshots**
![Screenshot_20220715-151503 (1)](https://user-images.githubusercontent.com/82892873/181037528-19e302d3-b72f-4da0-a16f-a758f0fc6249.png)
![IMG_0054 (1)](https://user-images.githubusercontent.com/82892873/181037545-633cdfd8-3d50-46fd-a2d5-921c3b3e24ed.PNG)
![IMG_0053 (1)](https://user-images.githubusercontent.com/82892873/181037560-27d6353b-de7e-4b6f-9bc0-23a0009edb5c.PNG)
**+ any other area horizontal scroll is seen.** 
Smartphone (please complete the following information):

Additional context
Examples of controls on another site:
https://user-images.githubusercontent.com/82892873/181037105-777c6335-5bcf-4976-9ee6-4419919e57a2.mp4
https://user-images.githubusercontent.com/82892873/181037139-e3d812fc-b5a4-4090-bf5d-cc763f6041ec.mp4

**Comment by TodorBachvarov** [comment]
Fixed in version 4.3.4.

**Comment by acovher2** [comment]
The verbiage pronounced by screen reader is a bit confusing. Since this is progressing through twitter cards and not actually through pages can we please change "next page" and "Previous page" to be " Next card" and "previous card" respectively?

**Comment by acovher2** [comment]
Awaiting version 4.3.4 The expected changed are not present yet. 
Expected behavior
There should be Previous/Next controls for any horizontally scrolling widget.
These should not be located on each individual card.
These Previous/next controls also need disabled/enabled states, as well as Pressed and Focus states.
The controls should meet color contrast ratio

**Comment by acovher2** [comment]
In version 4.2.52, The Daily Illini widget and the Twitter one both still have the controls on the cards rather than below or separately as seen in Al Groups, All Notifications, Campus Reminders and others. Why are these not the same? This would be best to be consistent and following these accessibility guidelines:
-These should not be located on each individual card.
-These Previous/next controls also need disabled/enabled states, as well as Pressed and Focus states.
-The controls should meet color contrast ratio

Currently, they are still not meeting those three. 
-Inconsistently inside the cards and below in different widgets.
-There is a removal of the next arrow when user is unable to proceed, however this is not the same as styling a disabled state from enabled because the assistive technology is not read the dynamic change of the button removal. Please leave the buttons and style and label them.

**Comment by acovher2** [comment]
https://user-images.githubusercontent.com/82892873/224024162-ee90f4eb-8deb-4e5d-854a-5eb3b8fa3bea.mp4

Video of current behavior with TalkBack.

**Comment by acovher2** [comment]
@pmarkhennessy Any updates on this issue? I still see same behavior in 6.0.6. 
@knguyenjahiel

**Comment by acovher2** [comment]
So the above defect, I was trying to make this better if it was a NECESSARY evil. However with the new Title II ADA changes, we need to be compliant and the above suggestions are not the way to do it. 

**Update about the impact of the violation:_** 
**Motor Impairments:**
**Challenges**: Users with conditions like arthritis, tremors, or limited dexterity may struggle with precise horizontal swipes. This can make it difficult for them to navigate through side-scrolling widgets.
**Impact**: These users might accidentally swipe too far or not far enough, leading to frustration and a poor user experience.
**Screen Reader Compatibility:**
**Challenges**: Screen readers often have difficulty detecting and announcing content that is off-screen in a side-scrolling widget.
**Impact**: Users who rely on screen readers may miss important information or options that are not immediately visible, leading to incomplete or confusing experiences.
**Cognitive Load:**
**Challenges**: Side-scrolling requires users to remember the position of items and navigate back and forth, which can be mentally taxing.
**Impact**: This can be particularly challenging for users with cognitive disabilities, such as memory impairments or attention disorders, making it harder for them to use the app effectively.
**Consistency:**
**Challenges**: Vertical scrolling is a more common and expected interaction pattern on mobile devices. Users are generally more familiar with this method of navigation.
**Impact**: Introducing side-scrolling can disrupt the user experience, making the app less intuitive and harder to navigate, especially for users who rely on predictable patterns.
**WCAG Guidelines:**
**Challenges**: The Web Content Accessibility Guidelines (WCAG) emphasize that content should be operable and understandable. Side-scrolling can violate these principles by making content harder to access and understand.
**Impact**: Non-compliance with WCAG can lead to legal issues and alienate users with disabilities. Ensuring vertical scrolling aligns better with accessibility standards and provides a more inclusive experience.


So we should probably consider new design solutions rather than my previously suggested tweaks.

**Comment by acovher2** [comment]
@pmarkhennessy Due to the ADA title II changes and deadline of April 2026, I wanted to ensure this gets on the list and radar as we need to be compliant. This was on the PPT I sent you.

**Comment by vburgett** [comment]
@acovher2 since you flagged this issue, I have noticed horizontal scrolling in so many tools--iPhone  iOS, Netflix, Google Maps, etc. Are all these companies not ADA compliant or is there a way to achieve accessible horizontal scrolling?

**Comment by acovher2** [comment]
I can't speak for items I haven't evaluated, but typically they aren't compliant if only designed for sighted users. 

Previously, I had suggested above, something to indicate they are horizontally swiping cards, next/previous buttons with enabled/disabled states appropriately ARIA labeled for Screen reader users so that they could tell when they had reached end of cards etc.  This also needs to account for a certain tab order in that you would not want the user to be hearing only "Next" or "Previous" you'd also want to go to the title of the card, give the user an option to select it (same as visual users get) and make the information announced to screen reader users equivalent to that the sighted user would see on the card.  

In light of the newer information about the affects of the violation, from September 10th, I'm not entirely sure which solutions to suggest. I would like to take this up the ladder so to speak and talk about impacts, and speak to my leadership about this. I know what you are saying about the popularity of this and use in apps, as well as even carousels on webpages to be honest. So, please stay tuned and I'll let you know what they say.

**Comment by acovher2** [comment]
Including a video of current behavior with TalkBack in 6.1.30 so you can see what I mean about tab order. And the next/previous buttons were never labeled or given enabled or disabled states. Additionally, there seems to be a new fun bug where it repeats the card name while swiping to the next and in between cards gets focus for some reason.... 

https://github.com/user-attachments/assets/60bdec1a-3870-415f-a74a-92228fd5d6c6

**Comment by acovher2** [comment]
@TodorBachvarov I'm attaching a video of this working well on iOS 18.3.2 iPhone 14+. I will open a separate bug for the new issue that is remaining on Android.  With the functionality of Screen reader gesturing included now, I am going to ask my leadership if we will be in compliance with that alone or if the arrows still need to have enabled and disabled states etc. 

https://github.com/user-attachments/assets/239545c8-5897-45f9-abdf-29f9b3977f90

**Comment by acovher2** [comment]
FYI this is how Dev 6.1.66 Android version is working with TalkBack, I'm going to open a bug about the focus shifting. there seems to be an extra spot between cards. 

https://github.com/user-attachments/assets/c4469c1c-b623-4dab-9bee-d67b97098347

**Comment by acovher2** [comment]
@TodorBachvarov This is much better experience in my opinion. 
**Improvements Verified:**
-There is a sound between card when swipe gesturing through
-Android focus issue resolved, no extra items between cards or extra sounds, consistently hearing 3 solid sounds between each card gesture and clear Category, Title, Image (they need alternative text, but that's another feature request #4836 .) 
-Next/Previous is disabled and receiving focus from screen reader
---There is one thing about this I don't know as I'm not a native user, I think it is easy enough to me to swipe gesture but I wonder if focus needs brought to the Next arrow first so that they have an option to know there is more content cycling or if they just indefinitely swipe until they reach last card, then what is the use of switching to the arrows receiving focus after? Perhaps to toggle back and find a certain card. 
-Either way you gesture or use arrows, it is much more screen reader friendly now. 

**Remaining issue with this ( I believe this should be a separate bug ticket, but wanted to ask you....** 
- There is movement unexpected of the next and previous arrows up and down as the card sizes flutuate.... It is subtle but some users may have issues with it but I'm unclear how big of an issue this is. I'm going to run it by someone affected by this condition but they are on vacation until Tuesday. 

Pending what I find out Tuesday I'll close this as verified and open separate issues for movement of arrows and the web version issues a little later on after this release. 

The Web version has it's own issues with keyboard navigation, I will open those bugs separately.

**Comment by TodorBachvarov** [comment]
Hi @acovher2, 
Regarding "---There is one thing about this I don't know as I'm not a native user,..."
As I remember the Next/previous button were introduced as a workaround when there were many issues with the swipe gesture/controls and it's VO support. This was the alternative to navigate trough the content with the VO on. Since the VO Swipe gesture support was improved in Flutter this remains the more natural way for navigation. 
"...then what is the use of switching to the arrows receiving focus after?"
The buttons are below the swipe content by design. 
It is recommended to avoid switching focus trough the elements because this may bring more confusion to the users. So I prefer not to change the focus order of the elements and the Next/Previous buttons.

"There is movement unexpected of the next and previous arrows up and down as the card sizes flutuate...."
If I'm understanding the issue correctly, 
The next/previous arrows are moving up/down when the content height is calculated and this is the behaviour of the swipe control that we use. So I'm not sure if we can improve this movement in any way since we want the swipe control to have different height depending on the visible content in it.

**Comment by acovher2** [comment]
Closing this one and tracking further discussion on [BUG] Accessibility - App - Side scrolling causing some animation-like effects #5259
